### PR TITLE
Table column resizing aria docs

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Looks at the column resizing docs for useTableState and useTable. In addition, try the styled example out here: https://codesandbox.io/s/objective-cherry-g837on?file=/src/App.tsx

## 🧢 Your Project:

RSP

### CODE REVIEW

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:248]: outline wasn't looking great in the resize example since it wasn't inset, replaced it with `boxShadow`. Did so for these pre-existing examples as well to reduce the diff between the this and the column resize examples

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:526]: Open to opinions. The changes made can be argued to be a bit extensive/excessive for an aria example but what I have here is what I felt to be a happy middle ground. Will be pointing out gotchas/places for alternative implementation here and there which I'd like opinions on

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:551]: Two possible strategies that I considered here. One is the current which is to make the table scrollable + sticky column headers (exactly how the aria example we have in the storybook is). Unfortunately this makes scrollIntoView iffy since it doesn't account for the sticky column headers taking room since it is floating on top of the scroll region.

Second strategy is to follow how we sync header scroll position in TableView. This felt a bit more complex for an aria example, but happy to implement if people think its ok

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:578]: this hard coded width  means the example is a bit limited when it comes to accepting table widths from the user such as `width: auto` or `maxWidth`/`minWidth`, etc. However the alternative is to use a resizeObserver which felt too much for the aria example (and had some finicky behavior will have to poke around)

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:732]: This is actually a bug (?) in our TableView, if you hover over the column headers and then use left/right to navigate through the columns, you'll actually land on the resizer and enter resizing mode. It also feels like a bit of a strange restriction to require the user to hide the resizer to avoid grid keyboard navigation issues

Did some thinking about this, maybe it would be nice to make an arbitrary element excluded from the FocusableTreewalker, open to ideas

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:747]: This is quite the change to the aria example's TableColumnHeader but IMO is good to have to keep interaction handling for all interaction types. Could be reduced if we wanna omit sorting handling but IMO its good to keep the previous functionality from the original aria example

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:948]: another interesting bug (?) that I encountered. Our MenuTrigger avoids this problem since the `onKeyDown` from `useMenuTrigger` is sent via context and thus `useFocusable` won't automatically `stopPropagation` on it via `useKeyboard` https://github.com/adobe/react-spectrum/blob/8e57eaeda9112ea0783432409b8e6f9416d2fb28/packages/%40react-aria/focus/src/useFocusable.tsx#L74-L76

The aria implementation of a MenuButton runs into this problem if it doesn't do something similar. Maybe add `continuePropagation` to useMenuTrigger's onKeydown? Feels too specific though

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:1280]: Maybe I'll include more examples (for example with sort)? Don't think we wanna include a resizable columns table in every example in the Usage section though, open to opinions here

File: packages/@react-stately/table/docs/useTableColumnResizeState.mdx
Comment (LFDanLu) [inline packages/@react-stately/table/docs/useTableColumnResizeState.mdx:22]: Feels more discoverable to have this as a separate page. `useTableColumnResize` feels ok to have in `useTable` docs since it modifies the original implementation and thus needs context

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:1284]: Hrm, experience w/ the resizer is broken in Safari + iffy in VO and Talkback, digging

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:578]: this hard coded width means the example is a bit limited when it comes to accepting table widths from the user such as width: auto or maxWidth/minWidth, etc. However the alternative is to use a resizeObserver which felt too much for the aria example (and had some finicky behavior will have to poke around)

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:732]: IMO that would depend if we'd want to make that treewalker filtering/excluding public api or for internal use only. Maybe it could be a prop/callback like you said, maybe we provide it through a context since useGridCell is a bit deep sometimes, maybe we have a specific DOM attribute we look for and auto filter elements that have that attribute (heh)

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:747]: Sounds good, I'll go ahead and make the docs example simple and do a tailwind example with all the bells and whistles

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:952]: makes sense, that would conform with our philosophy of stopping propagation for events in general. Bit unfortunate that this is a "gotcha" for people implementing this in their tables. 

If we are amenable to easing up the propagation stuff, I could see an argument to continue propagation for keydowns we don't explicitly handle maybe?

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:905]: orange sounds good, I'll go ahead with that

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:591]: I'll keep it as `display: block` and add a comment here so the changes between this and the previous implementation are as minimal as possible

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1101]: Maybe I'll rename MenuButton to MenuTrigger, MenuButton was the naming we have in the useMenu docs heh

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:952]: i think that would be fine, no reason to block keys we don't handle

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:1412]: Styled example in progress here: https://codesandbox.io/s/agitated-darwin-uki38i?file=/src/Table.tsx. I think the nightly is causing issues due to dupe packages, still need to investigate fully

File: packages/@react-aria/table/docs/useTable.mdx
Comment (reidbarber) [inline packages/@react-aria/table/docs/useTable.mdx:1415]: After pressing the 'Resize column' option in the menu, focus briefly goes back to the column header. Not a huge deal.

Comment (reidbarber) [reply packages/@react-aria/table/docs/useTable.mdx:1415]: Also, when dragging via mouse, the dragged column header will have a focus ring for a bit.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (reidbarber) [inline packages/@react-aria/table/docs/useTable.mdx:1190]: Stately links don't work in the build. I think they will still work in prod, but maybe safer to use relative path like `[useTreeState](../react-stately/useTreeState.html)`?

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1415]: I believe this is because it is using the button's native focus ring. I can override it, but was a bit wary of going too far with the custom styling

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [inline packages/@react-aria/table/docs/useTable.mdx:982]: qq, we no longer support sorting here, can we actually do away with the menu entirely and just have clicking on the column header go into resize mode for this example?

the tailwind one supports sorting and shows how they can work together.

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:985]: I considered doing that but it felt rather non-standard and a bit divergent from the tailwind example. I'd have to add aria-labeling/message to let screenreader users know that pressing on the header would trigger resizing and demonstrate overriding handlers from `useTableColumnHeader` in the off chance sorting was enabled as well. 

Personally I like having the menu here since it serves as a nice step 1.5 for people to consider what to do if they wanted additional actions on the TableColumnHeader. Happy to discuss if we still feel that the menu is over complicating things here

Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:985]: I'm fine with it as is.
We should update the sentence this is commenting on since the column header in this example is not reserved for sorting though.

**Review by snowystinger — APPROVED** [review]

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [inline packages/@react-aria/table/docs/useTable.mdx:830]: Flipped back to collapsing the whole code sections blocks, better experience on mobile IMO

**Review by reidbarber — APPROVED** [review]
LGTM

File: packages/@react-aria/table/docs/useTable.mdx
Comment (devongovett) [reply packages/@react-aria/table/docs/useTable.mdx:830]: I would expand them. The code is sorta the point of the docs I think. Most people will be looking at this on desktop as well.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (devongovett) [inline packages/@react-aria/table/docs/useTable.mdx:1400]: hmm a lot of the changes we are highlighting are purely style changes. Wonder if there is a way of simplifying this. Also, why does the table have a flex layout rather than table?

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:1004]: The issue with this was grid navigation. When arrow key navigating, you couldn't move between column headers. Instead, it'd start resizing.

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1004]: https://github.com/adobe/react-spectrum/pull/3982#discussion_r1091336950 alludes to this issue. When using the arrow keys to navigate between columns, you'll actually move focus to the resizer instead which means you are now in resize mode.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:1400]: display table has a lot of issues with respecting designated column sizes and cannot be horizontally scrolled.
For example, if you specify all columns with a min of 50px and then resize one column so it's huge and would cause horizontal scrolling, all the other columns will shrink to be smaller than the min specified.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1157]: makes sense, I'll open a separate PR

File: packages/@react-aria/table/docs/useTable.mdx
Comment (devongovett) [reply packages/@react-aria/table/docs/useTable.mdx:990]: hmm maybe we should have a mode where only the up/down arrows change the size rather than left/right? the menu solution is cool, but kinda spectrum design specific. I could see this being a frequent request.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:1372]: probably, but then you don't get the sticky headers if I recall correctly

File: packages/@react-aria/table/docs/useTable.mdx
Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:990]: that sounds reasonable, hadn't thought to divide up the actions like that

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:990]: I'll see if I can put a PR up for this as

**Review by snowystinger — APPROVED** [review]

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:990]: Ran into a snag trying to implement this. Since we `setKeyboardNavigationDisabled(true)` when the resizer is focused, I need to re-enable it when the user presses the left/right arrow keys. However, `useGridCell` `onKeyDownCapture` triggers before the `onKeyDownCapture` I setup in `useTableColumnResizing`, meaning we don't move focus back to the column header when on `ArrowLeft`.

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:990]: Similar issue happens if I get rid of `setKeyboardNavigationDisabled` entirely instead. Left/right arrow keys can be made to move between the resizer and the column but up/down arrows keys are then handled by `useGridCell` and results in focus moving from the resizer to the cell below.

Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:990]: ugh, yeah, that wouldn't be separable. keyboard grid navigation is either all or nothing... there's no way to distinguish between the two. we'd need to introduce a new prop or something and even then I'm not sure we could get ahead of it, as you mentioned you ran into a problem getting ahead of the capture events

Comment (snowystinger) [reply packages/@react-aria/table/docs/useTable.mdx:990]: could we change it from onFocus to on 'Return/Enter' pressed? maybe that could be the new prop?

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:990]: @devongovett any opinions on the above? Right now these are the options that I can see:
- make it so that "editMode" is enabled when the user presses Enter when focused on the resizer
  - perhaps non-intuitive?
- change the aria example so pressing enter on the column header enters resize mode
  - this doesn't solve grid navigation bit but it gets rid of the menu

Comment (devongovett) [reply packages/@react-aria/table/docs/useTable.mdx:990]: Either would be ok I think. Just need to make sure it's obvious somehow. Definitely would need an aria-label on the column header if that's how we go, but even for non-screen reader users it might not be obvious. Enter on the resizer itself is a little more I guess.

Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:990]: pushed code changes for pressing enter on the column header to enter resize mode (note that I haven't updated the docs copy to reflect the new example code). I'll open a separate branch for the "editMode" updates since that will require changes to the hooks themselves

File: packages/@react-aria/table/stories/example-docs.tsx
Comment (LFDanLu) [inline packages/@react-aria/table/stories/example-docs.tsx:120]: todo: don't forget to remove this

Comment (LFDanLu) [reply packages/@react-aria/table/stories/example-docs.tsx:120]: also make resizer no longer a forward ref

**Comment by LFDanLu** [comment]
TODO: wait till https://github.com/adobe/react-spectrum/pull/4077 is released before merging this. Code changes that exist in this pull are the same as in that pull

**Review by reidbarber — APPROVED** [review]
LGTM

File: packages/@react-aria/table/docs/useTable.mdx
Comment (devongovett) [inline packages/@react-aria/table/docs/useTable.mdx:1050]: wondering if we should even expose the `widths` directly from the hook. There is already a `getColumnWidth` method as well, which we could use here.

File: packages/@react-aria/table/docs/useTable.mdx
Comment (devongovett) [inline packages/@react-aria/table/docs/useTable.mdx:1135]: hmm maybe this prop should be called 'aria-label' so it's clear what it's for?

**Review by devongovett — COMMENTED** [review]
A few small suggestions

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1053]: It needs to be some kind of focusable element since focus will move to the [first focusable child](https://github.com/adobe/react-spectrum/blob/a16e8474ed620c8898b1c8133ad5ec796b25a773/packages/%40react-aria/table/src/useTableColumnHeader.ts#L48) due to `focusMode: child`. If I don't make the header content focusable, then focus will jump to the resizer. 

A button does feel odd since there isn't a press action but the other focusable elements don't feel appropriate either

File: packages/@react-aria/table/docs/useTable.mdx
Comment (LFDanLu) [reply packages/@react-aria/table/docs/useTable.mdx:1050]: makes sense, I can give that a try

**Comment by devongovett** [comment]
I also tried out an implementation that doesn't require overriding the table layout with flex and it removed a lot of the basic style overrides allowing us to reuse the components defined earlier. What do you think? https://github.com/adobe/react-spectrum/pull/4253

**Comment by LFDanLu** [comment]
@devongovett urgh, I wanted to preserve the sticky column headers but I suppose most people should opt to follow the styled example's approach anyways since having sorting + resizing is probably the most common use case.

**Comment by LFDanLu** [comment]
TODO: I'm updating the example to fit the changes we made to the resize hooks. Will update the codesandbox as well when the nightly gets bumped/we release the codechanges

**Comment by LFDanLu** [comment]
TODO: update sandbox example so it uses input styles from useTableColumnResize hook instead of VisuallyHidden, need to wait for nightly/release for that

**Review by snowystinger — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]

**Review by snowystinger — APPROVED** [review]

**Pull request merged**
